### PR TITLE
Added support for reporting telemetry from Azure CNI

### PR DESCRIPTION
**Pull request by tamilmani1989**
**What this PR does / why we need it**:
Added support for reporting telemetry from Azure CNI. This features gives statistics about CNI like how many nodes running CNI, how many failures in ADD/DEL calls. Apart from this, it also reports useful messages. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: cni/network/network.go [hunk 1]
```diff
@@ -589,6 +601,14 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
 
 	log.Printf("[cni-net] Read network configuration %+v.", nwCfg)
 
+	plugin.report.OperationType = "DEL"
+	if nwCfg.MultiTenancy {
+		plugin.report.Context = "AzureCNIMultitenancy"
+	}
+	plugin.report.SubContext = fmt.Sprintf("%+v", nwCfg)
+	plugin.report.BridgeDetails.NetworkMode = nwCfg.Mode
+	plugin.report.InterfaceDetails.SecondaryCAUsedCount = plugin.nm.GetNumberOfEndpoints("", nwCfg.Name)
```
Comment (ninzavivek) [inline cni/network/network.go]: could these lines be encapsulated in one function, seems like lots of repetitive code.

File: cni/network/plugin/main.go [hunk 2]
```diff
@@ -155,36 +154,61 @@ func main() {
...
+			BridgeDetails:    telemetry.BridgeInfo{},
 		},
 	}
 
-	reportManager.GetHostMetadata()
-	reportManager.Report.(*telemetry.CNIReport).GetReport(pluginName, config.Version, ipamQueryURL)
+	tb := telemetry.NewTelemetryBuffer()
+	attempt := 0
+
+	for attempt < 2 {
```
Comment (jaer-tsun) [inline cni/network/plugin/main.go]: for attempt := 0; attempt < 2; ++attempt {...}

File: cni/network/plugin/main.go [hunk 3]
```diff
@@ -155,36 +154,61 @@ func main() {
...
+			log.Printf("Connected to telemetry service")
+			break
+		}
+
+		attempt++
+	}
+
+	t := time.Now()
+	reportManager.Report.(*telemetry.CNIReport).Timestamp = t.Format("2006-01-02 15:04:05")
+
```
Comment (jaer-tsun) [inline cni/network/plugin/main.go]: remove space

File: cni/network/plugin/main.go [hunk 4]
```diff
@@ -155,36 +154,61 @@ func main() {
...
+	reportManager.Report.(*telemetry.CNIReport).GetReport(pluginName, version, ipamQueryURL)
```
Comment (jaer-tsun) [inline cni/network/plugin/main.go]: remove space

File: cni/network/plugin/main.go [hunk 5]
```diff
@@ -155,36 +154,61 @@ func main() {
...
 		if err != nil {
 			log.Printf("SendReport failed due to %v", err)
 		} else {
-			markSendReport(reportManager)
+			markSendReport(reportManager, tb)
 		}
 	}
 
+	startTime := time.Now().UnixNano() / int64(time.Millisecond)
+
```
Comment (jaer-tsun) [inline cni/network/plugin/main.go]: remove space

File: common/utils.go [hunk 6]
```diff
@@ -103,3 +103,27 @@ func GetInterfaceSubnetWithSpecificIp(ipAddr string) *net.IPNet {
...
+			os.Stdin,
+			nil,
+			nil,
+		},
+	}
+
+	args := []string{path}
+	process, err := os.StartProcess(path, args, &attr)
+	if err == nil {
+
```
Comment (jaer-tsun) [inline common/utils.go]: remove space

File: common/utils.go [hunk 7]
```diff
@@ -103,3 +103,27 @@ func GetInterfaceSubnetWithSpecificIp(ipAddr string) *net.IPNet {
...
+		// It is not clear from docs, but Realease actually detaches the process
```
Comment (jaer-tsun) [inline common/utils.go]: Release*

File: common/utils.go [hunk 8]
```diff
@@ -103,3 +103,27 @@ func GetInterfaceSubnetWithSpecificIp(ipAddr string) *net.IPNet {
...
+		if err = process.Release(); err != nil {
```
Comment (jaer-tsun) [inline common/utils.go]: return process.Release()

File: platform/os_linux.go [hunk 9]
```diff
@@ -94,3 +94,8 @@ func SetOutboundSNAT(subnet string) error {
 func ClearNetworkConfiguration() (bool, error) {
 	return false, nil
 }
+
+func KillProcessByName(processName string) {
```
Comment (jaer-tsun) [inline platform/os_linux.go]: don't we want to return the error from ExecuteCommand?

File: telemetry/cnstelemetry.go [hunk 10]
```diff
@@ -23,14 +22,20 @@ const (
...
+		if err := telemetryBuffer.Connect(); err != nil {
+			log.Printf("[Telemetry] Failed to establish telemetry manager connection.")
+			time.Sleep(time.Minute * 1)
+			goto CONNECT
+		}
+
+		go telemetryBuffer.BufferAndPushData(time.Duration(0))
 
-		heartbeat := time.NewTicker(time.Minute * 30).C
+		heartbeat := time.NewTicker(time.Minute * 1).C
```
Comment (jaer-tsun) [inline telemetry/cnstelemetry.go]: do we want 1 min heartbeats?

File: telemetry/telemetrybuffer.go [hunk 11]
```diff
@@ -205,15 +227,35 @@ func (tb *TelemetryBuffer) sendToHost() error {
 
 // push - push the report (x) to corresponding slice
 func (pl *Payload) push(x interface{}) {
+
```
Comment (jaer-tsun) [inline telemetry/telemetrybuffer.go]: remove space

File: cni/network/plugin/main.go (hunk 5 above)
Comment (tamilmani1989) [reply cni/network/plugin/main.go]: this is fine.

File: telemetry/cnstelemetry.go (hunk 10 above)
Comment (tamilmani1989) [reply telemetry/cnstelemetry.go]: oops..i forget to revert this change. Thanks for pointing out

File: cni/network/network.go [hunk 12]
```diff
@@ -156,6 +163,19 @@ func (plugin *netPlugin) getPodInfo(args string) (string, string, error) {
 	return k8sPodName, k8sNamespace, nil
 }
 
+func (plugin *netPlugin) setCNIReportDetails(nwCfg *cni.NetworkConfig, opType string, msg string) {
+	plugin.report.OperationType = opType
```
Comment (jaer-tsun) [inline cni/network/network.go]: not important, but it'd be cleaner if you move this line under the if statement

File: telemetry/telemetry_windows.go [hunk 13]
```diff
@@ -3,7 +3,17 @@
...
-import "runtime"
+import (
+	"runtime"
+	"strings"
+
+	"github.com/Azure/azure-container-networking/platform"
+)
+
+const (
+	delimiter  = "\r\n"
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply telemetry/telemetry_windows.go]: not that i know of

File: telemetry/telemetrybuffer_linux.go [hunk 14]
```diff
@@ -10,7 +10,10 @@ import (
 )
 
 const (
-	fdTemplate = "/tmp/%s.sock"
+	fdTemplate                  = "/tmp/%s.sock"
+	telemetryServiceProcessName = "azure-vnet-telemetry"
+	cniInstallDir               = "/opt/cni/bin"
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply telemetry/telemetrybuffer_linux.go]: this is telemetry specific..shouldn;t be in plaform package

File: common/utils.go [hunk 15]
```diff
@@ -103,3 +103,27 @@ func GetInterfaceSubnetWithSpecificIp(ipAddr string) *net.IPNet {
...
+			return err
+		}
+
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply common/utils.go]: thats the standard

**Pull request merged**